// lingo3d/lib/display/core/SimpleObjectManager/index.js
import { rad2Deg, deg2Rad, distance3d } from "@lincode/math";
import { Matrix3, Vector3 } from "three";
import { clickSet, mouseDownSet, mouseOutSet, mouseMoveSet, mouseOverSet, mouseUpSet } from "./raycast";
import { quaternion, ray, vector3, vector3_, vector3_1, vector3_half } from "../../utils/reusables";
import { forceGet } from "@lincode/utils";
import { OBB } from "three/examples/jsm/math/OBB";
import { scaleDown, scaleUp } from "../../../engine/constants";
import { addBloom, deleteBloom } from "../../../engine/render/effectComposer/selectiveBloomPass/renderSelectiveBloom";
import worldToClient from "../../utils/worldToClient";
import { Cancellable } from "@lincode/promiselikes";
import { point2Vec, vec2Point } from "../../utils/vec2Point";
import PhysicsItem from "./PhysicsItem";
import { physicsContactBodies, physicsContactMap } from "../../../physics/physicsLoop";
import { addSSR, deleteSSR } from "../../../engine/render/effectComposer/ssrPass";
const idMap = new Map();
const thisOBB = new OBB();
const targetOBB = new OBB();
const makeSet = () => new Set();
const ptDistCache = new WeakMap();
const distance3dCached = (pt, vecSelf) => {
    const cached = ptDistCache.get(pt);
    if (cached)
        return cached;
    const result = distance3d(pt.x, pt.y, pt.z, vecSelf.x * scaleUp, vecSelf.y * scaleUp, vecSelf.z * scaleUp);
    ptDistCache.set(pt, result);
    return result;
};
export default class SimpleObjectManager extends PhysicsItem {
    constructor(object3d) {
        super();
        this.object3d = object3d;
        this.outerObject3d = object3d;
        this.initOuterObject3d();
    }
    dispose() {
        if (this.done)
            return this;
        super.dispose();
        this._id !== undefined && idMap.get(this._id).delete(this);
        deleteSSR(this.object3d);
        return this;
    }
    find(name) {
        var _a;
        var _b;
        const child = this.outerObject3d.getObjectByName(name);
        return child && ((_a = (_b = child.userData).manager) !== null && _a !== void 0 ? _a : (_b.manager = new SimpleObjectManager(child)));
    }
    findAll(name) {
        const result = [];
        this.outerObject3d.traverse(child => {
            var _a;
            var _b;
            child.name === name && result.push((_a = (_b = child.userData).manager) !== null && _a !== void 0 ? _a : (_b.manager = new SimpleObjectManager(child)));
        });
        return result;
    }
    get onClick() {
        return this._onClick;
    }
    set onClick(cb) {
        var _a;
        (_a = this.clickHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onClick = cb;
        if (!cb)
            return;
        clickSet.add(this.object3d);
        this.clickHandle = this.watch(new Cancellable(() => clickSet.delete(this.object3d)));
    }
    get onMouseDown() {
        return this._onMouseDown;
    }
    set onMouseDown(cb) {
        var _a;
        (_a = this.mouseDownHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onMouseDown = cb;
        if (!cb)
            return;
        mouseDownSet.add(this.object3d);
        this.mouseDownHandle = this.watch(new Cancellable(() => mouseDownSet.delete(this.object3d)));
    }
    get onMouseUp() {
        return this._onMouseUp;
    }
    set onMouseUp(cb) {
        var _a;
        (_a = this.mouseUpHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onMouseUp = cb;
        if (!cb)
            return;
        mouseUpSet.add(this.object3d);
        this.mouseUpHandle = this.watch(new Cancellable(() => mouseUpSet.delete(this.object3d)));
    }
    get onMouseOver() {
        return this._onMouseOver;
    }
    set onMouseOver(cb) {
        var _a;
        (_a = this.mouseOverHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onMouseOver = cb;
        if (!cb)
            return;
        mouseOverSet.add(this.object3d);
        this.mouseOverHandle = this.watch(new Cancellable(() => mouseOverSet.delete(this.object3d)));
    }
    get onMouseOut() {
        return this._onMouseOut;
    }
    set onMouseOut(cb) {
        var _a;
        (_a = this.mouseOutHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onMouseOut = cb;
        if (!cb)
            return;
        mouseOutSet.add(this.object3d);
        this.mouseOutHandle = this.watch(new Cancellable(() => mouseOutSet.delete(this.object3d)));
    }
    get onMouseMove() {
        return this._onMouseMove;
    }
    set onMouseMove(cb) {
        var _a;
        (_a = this.mouseMoveHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._onMouseMove = cb;
        if (!cb)
            return;
        mouseMoveSet.add(this.object3d);
        this.mouseMoveHandle = this.watch(new Cancellable(() => mouseOutSet.delete(this.object3d)));
    }
    get name() {
        return this.outerObject3d.name;
    }
    set name(val) {
        this.outerObject3d.name = val;
    }
    get id() {
        return this._id;
    }
    set id(val) {
        this._id !== undefined && idMap.get(this._id).delete(this);
        this._id = val;
        val !== undefined && forceGet(idMap, val, makeSet).add(this);
    }
    getWorldPosition() {
        return vec2Point(this.object3d.getWorldPosition(vector3_));
    }
    getRay() {
        return ray.set(this.object3d.getWorldPosition(vector3_), this.object3d.getWorldDirection(vector3));
    }
    rayAt(distance) {
        return vec2Point(this.getRay().at(distance * scaleDown, vector3));
    }
    rayIntersectsAt(target, maxDistance) {
        if (this.done)
            return undefined;
        if (target.done)
            return undefined;
        if (this === target)
            return undefined;
        targetOBB.set(target.object3d.getWorldPosition(new Vector3()), vector3_half, new Matrix3().setFromMatrix4(target.object3d.matrixWorld));
        const vec = targetOBB.intersectRay(this.getRay(), vector3);
        if (!vec)
            return;
        if (maxDistance) {
            const { x, y, z } = this.object3d.getWorldPosition(vector3_);
            if (distance3d(vec.x, vec.y, vec.z, x, y, z) * scaleUp > maxDistance)
                return;
        }
        return vec2Point(vec);
    }
    rayIntersects(target) {
        return !!this.rayIntersectsAt(target);
    }
    getRayIntersectionsAt(id, maxDistance) {
        var _a;
        const result = [];
        for (const target of (_a = idMap.get(id)) !== null && _a !== void 0 ? _a : []) {
            if (target === this)
                continue;
            const pt = this.rayIntersectsAt(target, maxDistance);
            pt && result.push([target, pt]);
        }
        this.object3d.getWorldPosition(vector3_);
        return result.sort((a, b) => {
            return distance3dCached(a[1], vector3_) - distance3dCached(b[1], vector3_);
        });
    }
    getRayIntersections(id, maxDistance) {
        return this.getRayIntersectionsAt(id, maxDistance).map(result => result[0]);
    }
    listenToRayIntersection(id, cb, maxDistance) {
        return this.loop(() => {
            for (const [target, pt] of this.getRayIntersectionsAt(id, maxDistance))
                cb(target, pt);
        });
    }
    intersects(target) {
        var _a, _b;
        if (this.done)
            return false;
        if (target.done)
            return false;
        if (this === target)
            return false;
        if (this.physicsBody && target.physicsBody) {
            physicsContactBodies.add(this.physicsBody);
            physicsContactBodies.add(target.physicsBody);
            return (((_a = physicsContactMap.get(this.physicsBody)) === null || _a === void 0 ? void 0 : _a.has(target.physicsBody)) ||
                ((_b = physicsContactMap.get(target.physicsBody)) === null || _b === void 0 ? void 0 : _b.has(this.physicsBody)) || false);
        }
        thisOBB.set(this.object3d.getWorldPosition(new Vector3()), vector3_1.clone(), new Matrix3());
        thisOBB.applyMatrix4(this.object3d.matrixWorld);
        targetOBB.set(target.object3d.getWorldPosition(new Vector3()), vector3_1.clone(), new Matrix3());
        targetOBB.applyMatrix4(target.object3d.matrixWorld);
        return thisOBB.intersectsOBB(targetOBB, 0);
    }
    getIntersections(id) {
        var _a;
        const result = [];
        for (const target of (_a = idMap.get(id)) !== null && _a !== void 0 ? _a : []) {
            if (target === this)
                continue;
            this.intersects(target) && result.push(target);
        }
        return result;
    }
    listenToIntersection(id, cb) {
        return this.loop(() => {
            for (const target of this.getIntersections(id))
                cb(target);
        });
    }
    get onIntersect() {
        return this._onIntersect;
    }
    set onIntersect(cb) {
        this._onIntersect = cb;
        if (this._onIntersectHandles)
            for (const handle of this._onIntersectHandles)
                handle.cancel();
        if (!cb || !this.intersectIDs)
            return;
        const handles = this._onIntersectHandles = [];
        for (const id of this.intersectIDs)
            handles.push(this.listenToIntersection(id, cb));
    }
    get clientX() {
        return worldToClient(this.object3d).x;
    }
    get clientY() {
        return worldToClient(this.object3d).y;
    }
    get width() {
        return this.object3d.scale.x * scaleUp;
    }
    set width(val) {
        this.object3d.scale.x = val * scaleDown;
    }
    get height() {
        return this.object3d.scale.y * scaleUp;
    }
    set height(val) {
        this.object3d.scale.y = val * scaleDown;
    }
    get depth() {
        return this.object3d.scale.z * scaleUp;
    }
    set depth(val) {
        this.object3d.scale.z = val * scaleDown;
    }
    get x() {
        return this.outerObject3d.position.x * scaleUp;
    }
    set x(val) {
        var _a;
        var _b;
        this.outerObject3d.position.x = val * scaleDown;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).position) !== null && _a !== void 0 ? _a : (_b.position = {})).x = true);
    }
    get y() {
        return this.outerObject3d.position.y * scaleUp;
    }
    set y(val) {
        var _a;
        var _b;
        this.outerObject3d.position.y = val * scaleDown;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).position) !== null && _a !== void 0 ? _a : (_b.position = {})).y = true);
    }
    get z() {
        return this.outerObject3d.position.z * scaleUp;
    }
    set z(val) {
        var _a;
        var _b;
        this.outerObject3d.position.z = val * scaleDown;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).position) !== null && _a !== void 0 ? _a : (_b.position = {})).z = true);
    }
    get scaleX() {
        return this.outerObject3d.scale.x;
    }
    set scaleX(val) {
        this.outerObject3d.scale.x = val;
    }
    get scaleY() {
        return this.outerObject3d.scale.y;
    }
    set scaleY(val) {
        this.outerObject3d.scale.y = val;
    }
    get scaleZ() {
        return this.outerObject3d.scale.z;
    }
    set scaleZ(val) {
        this.outerObject3d.scale.z = val;
    }
    get scale() {
        return this.scaleX;
    }
    set scale(val) {
        this.scaleX = val;
        this.scaleY = val;
        this.scaleZ = val;
    }
    get rotationX() {
        return this.outerObject3d.rotation.x * rad2Deg;
    }
    set rotationX(val) {
        var _a;
        var _b;
        this.outerObject3d.rotation.x = val * deg2Rad;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).rotation) !== null && _a !== void 0 ? _a : (_b.rotation = {})).x = true);
    }
    get rotationY() {
        return this.outerObject3d.rotation.y * rad2Deg;
    }
    set rotationY(val) {
        var _a;
        var _b;
        this.outerObject3d.rotation.y = val * deg2Rad;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).rotation) !== null && _a !== void 0 ? _a : (_b.rotation = {})).y = true);
    }
    get rotationZ() {
        return this.outerObject3d.rotation.z * rad2Deg;
    }
    set rotationZ(val) {
        var _a;
        var _b;
        this.outerObject3d.rotation.z = val * deg2Rad;
        this.physicsUpdate && (((_a = (_b = this.physicsUpdate).rotation) !== null && _a !== void 0 ? _a : (_b.rotation = {})).z = true);
    }
    get rotation() {
        return this.rotationZ;
    }
    set rotation(val) {
        this.rotationZ = val;
    }
    get reflection() {
        return !!this.object3d.userData.ssr;
    }
    set reflection(val) {
        val ? addSSR(this.object3d) : deleteSSR(this.object3d);
    }
    get bloom() {
        return !!this.outerObject3d.userData.bloom;
    }
    set bloom(val) {
        val ? addBloom(this.outerObject3d) : deleteBloom(this.outerObject3d);
    }
    get visible() {
        return this.outerObject3d.visible;
    }
    set visible(val) {
        this.outerObject3d.visible = val;
    }
    lookAt(target) {
        if ("object3d" in target)
            this.outerObject3d.lookAt(target.object3d.getWorldPosition(vector3));
        else
            this.outerObject3d.lookAt(point2Vec(target));
        this.physicsUpdateRotation();
    }
    translateX(val) {
        this.outerObject3d.translateX(val * scaleDown);
        this.physicsUpdatePosition();
    }
    translateY(val) {
        this.outerObject3d.translateY(val * scaleDown);
        this.physicsUpdatePosition();
    }
    translateZ(val) {
        this.outerObject3d.translateZ(val * scaleDown);
        this.physicsUpdatePosition();
    }
    rotateX(val) {
        this.outerObject3d.rotateX(val * deg2Rad);
        this.physicsUpdateRotation();
    }
    rotateY(val) {
        this.outerObject3d.rotateY(val * deg2Rad);
        this.physicsUpdateRotation();
    }
    rotateZ(val) {
        this.outerObject3d.rotateZ(val * deg2Rad);
        this.physicsUpdateRotation();
    }
    placeAt(object) {
        if ("object3d" in object) {
            this.outerObject3d.position.copy(object.object3d.getWorldPosition(vector3));
            this.outerObject3d.quaternion.copy(object.outerObject3d.getWorldQuaternion(quaternion));
        }
        else
            this.outerObject3d.position.set(object.x * scaleDown, object.y * scaleDown, object.z * scaleDown);
        this.physicsUpdatePosition();
        this.physicsUpdateRotation();
    }
    moveForward(distance) {
        if (distance === 0)
            return;
        if (this.physicsBody) {
            const direction = this.object3d.getWorldDirection(vector3);
            const d = -distance * 0.6;
            this.physicsBody.velocity.set(direction.x * d, this.physicsBody.velocity.y, direction.z * d);
            return;
        }
        vector3.setFromMatrixColumn(this.outerObject3d.matrix, 0);
        vector3.crossVectors(this.outerObject3d.up, vector3);
        this.outerObject3d.position.addScaledVector(vector3, distance * scaleDown);
        this.physicsUpdatePositionXZ();
    }
    moveRight(distance) {
        if (distance === 0)
            return;
        vector3.setFromMatrixColumn(this.outerObject3d.matrix, 0);
        this.outerObject3d.position.addScaledVector(vector3, distance * scaleDown);
        this.physicsUpdatePositionXZ();
    }
}
