// lingo3d/lib/display/cameras/ThirdPersonCamera.js
import Point3d from "../../api/Point3d";
import Camera from "./Camera";
export default class ThirdPersonCamera extends Camera {
    constructor() {
        super();
        this.mouseControlMode = 1;
        this.rotationY = 180;
        this.innerZ = 200;
    }
    get target() {
        return this._target;
    }
    set target(target) {
        var _a;
        (_a = this._loopHandle0) === null || _a === void 0 ? void 0 : _a.cancel();
        if (!target)
            return;
        let targetPoint = this.rayAt(10000);
        this._loopHandle0 = this.loop(() => {
            targetPoint = this.rayAt(10000);
            targetPoint.y += 2000;
            target.lookAt(new Point3d(targetPoint.x, 0, targetPoint.z));
            this.x = target.x;
            this.z = target.z;
            this.y = target.y;
        });
    }
    append(object) {
        var _a;
        (_a = this.outerObject3d.parent) === null || _a === void 0 ? void 0 : _a.add(object.outerObject3d);
        this.target = object;
    }
}
