// lingo3d/lib/api/settings.js
import { getFillWindow, setFillWindow } from "../states/useFillWindow";
import { getContainerSize, setContainerSize } from "../states/useContainerSize";
import { getPixelRatio, setPixelRatio } from "../states/usePixelRatio";
import { getResolutionScale, setResolutionScale } from "../states/useResolutionScale";
import { getViewportSize, setViewportSize } from "../states/useViewportSize";
import { getDefaultFog, setDefaultFog } from "../states/useDefaultFog";
import { getDefaultLight, setDefaultLight } from "../states/useDefaultLight";
import { getGridHelper, setGridHelper } from "../states/useGridHelper";
import { getCameraHelper, setCameraHelper } from "../states/useCameraHelper";
import { getLightHelper, setLightHelper } from "../states/useLightHelper";
import { getPerformance, setPerformance } from "../states/usePerformance";
import { getOrbitControls, setOrbitControls } from "../states/useOrbitControls";
export default {
    get fillWindow() {
        return getFillWindow();
    },
    set fillWindow(value) {
        setFillWindow(value);
    },
    get width() {
        return getViewportSize()[0];
    },
    set width(val) {
        setViewportSize([val, getViewportSize()[1]]);
    },
    get height() {
        return getViewportSize()[1];
    },
    set height(val) {
        setViewportSize([getViewportSize()[0], val]);
    },
    get containerWidth() {
        return getContainerSize()[0];
    },
    set containerWidth(val) {
        setContainerSize([val, getContainerSize()[1]]);
    },
    get containerHeight() {
        return getContainerSize()[1];
    },
    set containerHeight(val) {
        setContainerSize([getContainerSize()[0], val]);
    },
    get pixelRatio() {
        return getPixelRatio();
    },
    set pixelRatio(value) {
        setPixelRatio(value);
    },
    get resolutionScale() {
        return getResolutionScale();
    },
    set resolutionScale(value) {
        setResolutionScale(value);
    },
    get performance() {
        return getPerformance();
    },
    set performance(value) {
        setPerformance(value);
    },
    get gridHelper() {
        return getGridHelper();
    },
    set gridHelper(value) {
        setGridHelper(value);
    },
    get cameraHelper() {
        return getCameraHelper();
    },
    set cameraHelper(value) {
        setCameraHelper(value);
    },
    get lightHelper() {
        return getLightHelper();
    },
    set lightHelper(value) {
        setLightHelper(value);
    },
    get defaultFog() {
        return getDefaultFog();
    },
    set defaultFog(value) {
        setDefaultFog(value);
    },
    get defaultLight() {
        return getDefaultLight();
    },
    set defaultLight(value) {
        setDefaultLight(value);
    },
    get defaultOrbitControls() {
        return getOrbitControls();
    },
    set defaultOrbitControls(value) {
        setOrbitControls(value);
    },
    wasmPath: "https://www.lingo3d.com/wasm/"
};
