// @pinyinma/standardlib/lib/index.js
import { pull, splitFileName } from "@lincode/utils";
import store from "@lincode/reactivity";
export const $eWRhpRV$loadingCount = store(0);
export const print = (...args) => {
    console.log(...args);
};
export const random = (x, y) => Math.floor(Math.random() * ((y - x) + 1) + x);
export const abs = Math.abs;
export const atan2 = (y, x) => Math.atan2(y, x) * 180 / Math.PI;
export const $eWRhpRV$range = (start, end) => {
    const result = [];
    if (start < end)
        for (let i = start; i <= end; ++i)
            result.push(i);
    else
        for (let i = start; i >= end; --i)
            result.push(i);
    return result;
};
const preloadCache = new Set();
const [setLoadingCount, getLoadingCount] = $eWRhpRV$loadingCount;
export const $eWRhpRV$preloaders = new Map();
export const $eWRhpRV$preload = (url) => {
    var _a;
    if (preloadCache.has(url))
        return url;
    preloadCache.add(url);
    const extension = (_a = splitFileName(url)[1]) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    if (extension && $eWRhpRV$preloaders.has(extension)) {
        setLoadingCount(getLoadingCount() + 1);
        $eWRhpRV$preloaders.get(extension)(url).finally(() => setLoadingCount(getLoadingCount() - 1));
    }
    return url;
};
//@ts-ignore
Array.prototype.delete = function (item) {
    pull(this, item);
};
//@ts-ignore
Array.prototype.append = Array.prototype.push;
