// lingo3d/lib/display/utils/deserialize/schema.js
const objectManagerProps = {
    x: 0,
    y: 0,
    z: 0,
    width: 100,
    height: 100,
    depth: 100,
    scale: 1,
    scaleX: 1,
    scaleY: 1,
    scaleZ: 1,
    rotation: 0,
    rotationX: 0,
    rotationY: 0,
    rotationZ: 0,
    innerX: 0,
    innerY: 0,
    innerZ: 0,
    innerRotation: 0,
    innerRotationX: 0,
    innerRotationY: 0,
    innerRotationZ: 0,
    bloom: false,
    reflection: false,
    visible: true
};
const loadedProps = {
    ...objectManagerProps,
    src: ""
};
const texturedBasicProps = {
    color: "",
    vertexColors: true,
    fog: true,
    opacity: 0,
    texture: "",
    alphaMap: ""
};
const texturedStandardProps = {
    color: "",
    flatShading: true,
    wireframe: true,
    envMap: "",
    aoMap: "",
    aoMapIntensity: 0,
    bumpMap: "",
    bumpScale: 0,
    displacementMap: "",
    displacementScale: 0,
    displacementBias: 0,
    emissiveColor: "",
    emissiveMap: "",
    emissiveIntensity: 0,
    lightMap: "",
    lightMapIntensity: 0,
    metalnessMap: "",
    metalness: 0,
    roughnessMap: "",
    roughness: 0,
    normalMap: "",
    normalScale: 0,
    normalMapType: "objectSpace",
    refractionRatio: 0
};
const lightProps = {
    ...objectManagerProps,
    color: "",
    intensity: 0
};
const primitiveProps = {
    ...texturedBasicProps,
    ...texturedStandardProps,
    ...objectManagerProps
};
const group = objectManagerProps;
const model = {
    ...loadedProps,
    loadedScale: 1,
    loadedX: 0,
    loadedY: 0,
    loadedZ: 0
};
const svgMesh = {
    ...loadedProps,
    ...primitiveProps
};
const sprite = {
    ...texturedBasicProps,
    ...objectManagerProps
};
const reflector = {
    ...objectManagerProps,
    contrast: 0,
    blur: 0
};
const scene = loadedProps;
const ambientLight = lightProps;
const areaLight = {
    ...lightProps,
    power: 0
};
const directionalLight = lightProps;
const skyLight = {
    ...lightProps,
    groundColor: ""
};
const pointLight = {
    ...lightProps,
    decay: 0,
    distance: 0,
    power: 0
};
const spotLight = {
    ...pointLight,
    angle: 0,
    penumbra: 0
};
const camera = {
    ...objectManagerProps,
    fov: 0,
    zoom: 0,
    bokeh: true,
    bokehFocus: 0,
    bokehMaxBlur: 0,
    bokehAperture: 0
};
const circle = primitiveProps;
const cone = primitiveProps;
const cube = primitiveProps;
const cylinder = primitiveProps;
const octahedron = primitiveProps;
const plane = primitiveProps;
const sphere = primitiveProps;
const tetrahedron = primitiveProps;
const torus = primitiveProps;
export default {
    group,
    model,
    svgMesh,
    sprite,
    reflector,
    scene,
    ambientLight,
    areaLight,
    directionalLight,
    skyLight,
    pointLight,
    spotLight,
    camera,
    circle,
    cone,
    cube,
    cylinder,
    octahedron,
    plane,
    sphere,
    tetrahedron,
    torus
};
